### DISCUSSION TITLE
Feature Request: Passing Team ID via Named Argument

### DISCUSSION THREAD

**Discussion by AAllport** [spatie/laravel-permission#1993] [Q&A] [answered]
PHP 8 brings with it [named arguments](https://www.php.net/manual/en/functions.arguments.php#functions.named-arguments).

PHP7 is no longer under active support: https://www.php.net/supported-versions.php 
Laravel 9 plans to raise the minimum version to PHP8: https://laravel.com/docs/master/upgrade#updating-dependencies

Given the release of the new teams features (and the fact php7 is no longer under active development), this could be a good opportunity to expand the API by allowing developers to pass the team id as a named argument.
This would preserve backwards compatibility with the `$guardName` argument, whilst allowing developers to utilise the teams feature without continually calling `setPermissionsTeamId`

Example usage: In a multi-team app, with a stateless API, it may be desirable to explicitly check for permission during a model gate.
This feature would allow users to call the following:
```php
    public function create(User $user, Post $post)
    {
        $user->hasPermission('post.create',teamId: $post->team_id);
    }
```

Happy to implement this if a PR is likely to be merged!

**Comment by erikn69** [comment, marked as answer]
>$user->hasPermission('post.create',teamId: $post->team_id);

that will not work because `hasPermission` uses preload `permissions/roles` relations, and there is a `wherePivot` 

https://github.com/spatie/laravel-permission/blob/6a3ed627cee28a552b5176c172ae0abc5eb30925/src/Traits/HasPermissions.php#L60
https://github.com/spatie/laravel-permission/blob/6a3ed627cee28a552b5176c172ae0abc5eb30925/src/Traits/HasRoles.php#L57

**Reply by AAllport** [reply]
Ahh, didn't realise the library was preloading.
I'm guessing in the use-case of users being parts of multiple teams isn't quite covered?

Eg. In a schema where you have Teams, which have Posts (and other child entities of teams) and Users
You would expect to know the team the user is acting under as early as possible (as to be able to execute the preload)
Whereas, if we're developing an API where the user can choose to act on any teams they have permission to act on

Could we entertain the idea of allowing developers to opt-out of the pivot there (causing all of the permissions to be loaded)
This would allow the check for permissions to happen when the policy is called?

Follow-up thought... laravel-permissions isn't triggering an eager load is it?
As such, our use-case could be tackled by doing the setPermissionsTeamId() immediately before doing the permission check?

The alternative (unless I'm mistaken), is creating a bunch of permissions that are `post.create.<team_id>` and then doing checks on that? (Although, that feels a tad messy)

**Reply by AAllport** [reply]
Follow up thought, due to how the team system is working, there's not currently a way for another user (eg, an admin) to view the permissions that another user has for a specific team via `$user->roles` (without some finagling)?

Happy to draw out some diagrams if that illustrates what I'm getting at, etc.

**Reply by erikn69** [reply]
>I'm guessing in the use-case of users being parts of multiple teams isn't quite covered?

The library was no made for that, and not even going to change for that

>Although, that feels a tad messy

a lot of messy(also the proposed ways) 

>Could we entertain the idea of allowing developers to opt-out of the pivot there

You can overwrite everything on your implementation

>As such, our use-case could be tackled by doing the setPermissionsTeamId() immediately before doing the permission check?
 
Yes

>Whereas, if we're developing an API where the user can choose to act on any teams they have permission to act on

I think like you need other library, or just make a relation on user with teams, and make a custom validation rule for validate if user has access to resource like `'team_id' => ['required', new CustomValidateTeam()],` where `CustomValidateTeam` uses `auth()->user()->teams` for validate [custom-validation-rules](https://laravel.com/docs/8.x/validation#custom-validation-rules)

**Reply by AAllport** [reply]
Thanks for following up with me!

We'll probably start with ensuring we set the team id moments before the check!
And then as we go forward, should we need to, we'll consider using a different permissions structure!

Thanks!

**Reply by erikn69** [reply]
> We'll probably start with ensuring we set the team id moments before the check!

You can make  a middleare for that, mark this as answered if you don't have more questions
